okteto/okteto
# DEV-690: Use right build context path to get the right diff and sha in git operations for Smart Builds

# Proposed changes

Fixes DEV-690

When images specify a parent folder as context, or if the command is executed from a parent folder and the build context a is grand-child folder, the calculation of the Smart Builds context wasn't being done correctly, because the path specified to git commands wasn't right.

Apart from that, in some cases, the build context is a relative path (when the build information is specified in an Okteto manifest), and other times it is an absolute path (when the build information comes from a compose file), so it was confusing the path passed to git commands.

With these changes, we now send always to git a local absolute path to the folder containing the image build context:
* If the build context is already an absolute path, we just pass that value to the git command, as it was already calculated properly
* If the build context is a relative path, we just join the current working directory + the build context. As the build context was already calculated based on the working directory, joining both of them, we have the right absolute path, so git always works with absolute routes to folders.

I added also a few debug commands, so we can see in the traces which commands are executed with git to calculate the build context.

## How to validate

Using the repository https://github.com/ifbyol/test-okteto and the branch `ifbyol/test-smart-build-contexts`, you could test all the different scenarios (each folder is an scenario to test). Without my changes, you can see how some of those cases don't work as expected, and after my changes, smart builds works as expected.

To validate it properly, it should be executed from the root of the repo, but also from other paths, to make sure we always specify the right path to git to do the Smart Build calculations

## CLI Quality Reminders 🔧

For both authors and reviewers:

- Scrutinize for potential regressions
- Ensure key automated tests are in place
- Build the CLI and test using the validation steps
- Assess Developer Experience impact (log messages, performances, etc)
- If too broad, consider breaking into smaller PRs
- Adhere to our [code style](https://github.com/okteto/okteto/blob/master/docs/code-style.md) and [code review](https://github.com/okteto/okteto/blob/master/docs/code-review.md) guidelines
